packages/react-mutation-mapper/src/component/track/ExonNumTrack.tsx

```diff
@@ -0,0 +1,160 @@
+> {
+    constructor(props: any) {
+        super(props);
+        makeObservable(this);
+    }
+
+    @computed get exonNumSpecs(): TrackItemSpec[] {
+        let currentTranscriptId = this.props.store.activeTranscript?.result;
+        let data: number[][] = [];
+        if (currentTranscriptId !== undefined) {
+            let currentTranscript = this.props.store.transcriptsByTranscriptId[
+                currentTranscriptId
+            ];
+            if (currentTranscript !== undefined) {
+                let exonArr = currentTranscript['exons'];
+                let utrArr = currentTranscript['utrs'];
+                if (utrArr === undefined) {
+                    utrArr = [];
+                }
+                let totLength = 0;
+                let exonLocList: number[][] = [];
+                // loop through exons
+                for (let i = 0; i < exonArr.length; i++) {
+                    let exonStart = exonArr[i].exonStart;
+                    let exonEnd = exonArr[i].exonEnd;
+                    let weDidIt = false;
+                    for (let j = 0; j < utrArr.length; j++) {
+                        let currentUtr = utrArr[j];
+                        // if utr start is within exon, add only translated length of exon to exonLocList
+                        if (
+                            exonStart <= currentUtr.start &&
+                            exonEnd >= currentUtr.start
+                        ) {
+                            weDidIt = true;
+                            let aaLength =
+                                (currentUtr.start -
+                                    exonStart +
+                                    (exonEnd - currentUtr.end)) /
+                                3;
+                            if (aaLength !== 0) {
+                                exonLocList.push([exonArr[i].rank, aaLength]);
+                                totLength += aaLength;
+                            }
+                            break;
+                        }
+                    }
+                    // if there are no utr start sites within exon
+                    if (!weDidIt) {
+                        let aaLength = (exonEnd - exonStart + 1) / 3;
+                        exonLocList.push([exonArr[i].rank, aaLength]);
+                        totLength += aaLength;
+                    }
+                }
+                exonLocList.sort((n1, n2) => n1[0] - n2[0]);
+                if (totLength !== this.props.proteinLength) {
+                    exonLocList[exonLocList.length - 1][1] =
+                        exonLocList[exonLocList.length - 1][1] - 1;
+                    totLength--;
+                }
+                let startOfExon = 0;
+                for (let i = 0; i < exonLocList.length; i++) {
+                    let exonLen = exonLocList[i][1];
+                    data.push([exonLocList[i][0], exonLen, startOfExon]);
+                    startOfExon += exonLen;
+                }
```

Comment:
You could consider making this a `map` call. Not sure I like it because of the side effect, but here's the gist of it:
```suggestion
            const data: ExonDatum[] = exonLocList.map(exonLocation => {
                const ret = {
                    location: exonLocation[0],
                    length: exonLocation[1],
                    start: startOfExon,
                };
                startOfExon += exonLocation[1];
                return ret;
            });
```